This directory contains source accompanying PLOS ONE manuscript " Computational characterization of modes of transcriptional regulation of nuclear receptor genes". 

List of files
-------------
1) count_hcne.pl 
2) plot_hcnematrix.R
3) plot_hcnecdf.R
4) Statisticaltest_enhancerenrichment_v1.R
5) Statisticaltest_enhancerenrichment_v2.R
6) plot_statisticalanalysis.R
7) fetch_significantreads.pl
8) plot_bubbleplot.R
9) plot_HMMstate.R
10) extract_coverage.R

Individual descriptions
-----------------------

1) This script count the HCNE densities around 2MB region of gene coordinates for single pair wise comparison. All the five  pairwise HCNEs were counted the same way.
Usage: Perl count_hcne.pl >outfilename  

2) This script generate HCNE matrix shown in figure 2 using the output file generated by above script for all the species.
Usage: R CMD BATCH  plot_hcnematrix.R

3) This	 script generate cumulative distribution of HCNE densities for each pair wise comparison  shown in figure 3. 
Usage: R CMD BATCH  plot_hcnecdf.R

4) This code generate data files for background distribution of H3k4me1 enrichment for the datasets used for statistical analysis for all selected genomic ranges.
Usage: R CMD BATCH Statisticaltest_enhancerenrichment_v1.R

5) This code generate data files for count of H3k4me1 reads for each of the datasets through random sampling for all selected genomic ranges.
Usage: R CMD BATCH Statisticaltest_enhancerenrichment_v2.R

6) This script generate statistical distribution of H3k4me1 read counts in each dataset shown in Figure 4.
Usage: R CMD BATCH plot_statisticalanalysis.R

7) This script counts the number of significant reads within 10kb region of gene co-ordinates for specific histone mark.
Usage: Perl fetch_significantreads.pl >outfilename

8) This script generate bubble plot shown in figure 5. The input files for the data were generated using the above script (fetch_significantreads.pl) for both histone marks.
Usage: R CMD BATCH plot_bubbleplot.R

9) This	 script generate bubble plot shown in figure 6.
Usage: R CMD BATCH plot_HMMstate.R

10) This code generate files for extracting coverage across selected genomic ranges and finally plot the coverage for H3k4me2 for Gm12878 cell line. The coverage for other histone marks were extracted the same way across all cell lines. 
Usage: R CMD BATCH extract_coverage.R
